File: frontend/src/types/summary.ts
import type { Transaction } from "./transaction";

export interface ValueAndChange {
	value: number;
	change: number;
}

export interface CategorySpending {
	category: string;
	amount: number;
}

export interface DateSpending {
	date: string;
	amount: number;
}

export interface MonthSpending {
	month: string;
	amount: number;
}

export interface CategorySpendingChange {
	category: string;
	amount: number;
	change: number;
	percentageChange: number;
	isNew: boolean;
}

export interface Merchant {
	name: string;
	category: string;
	amount: number;
}

export interface Summary {
	totalIncome: number;
	totalSpent: ValueAndChange;
	netSavings: ValueAndChange;
	savingsRate: ValueAndChange;
	lastMonthSpent: number;
	topCategory: CategorySpending;
	monthlySummary: string;
	topCategories: CategorySpending[];
	monthlyExpenses: DateSpending[];
	dailyTrend: DateSpending[];
	biggestChanges: CategorySpendingChange[];
	topMerchants: Merchant[];
	recentTransactions: Transaction[];
}

File: bff/internal/model/summary.go
package model

type ValueAndChange[T int | float64] struct {
	Value  T       `json:"value"`
	Change float64 `json:"change"`
}

type TransactionSummary struct {
	TotalIncome        int                      `json:"totalIncome"`
	TotalSpent         ValueAndChange[int]      `json:"totalSpent"`
	NetSavings         ValueAndChange[int]      `json:"netSavings"`
	SavingsRate        ValueAndChange[float64]  `json:"savingsRate"`
	LastMonthSpent     int                      `json:"lastMonthSpent"`
	TopCategory        *CategorySpending        `json:"topCategory"`
	MonthlySummary     string                   `json:"monthlySummary"`
	SpendingByCategory []CategorySpending       `json:"spendingByCategory"`
	MonthlyTrend       []DateSpending           `json:"monthlyTrend"`
	DailyTrend         []DateSpending           `json:"dailyTrend"`
	BiggestChanges     []CategorySpendingChange `json:"biggestChanges"`
	TopMerchants       []Merchant               `json:"topMerchants"`
	RecentTransactions []Transaction            `json:"recentTransactions"`
}
